components/common/src/ui.rs

```diff
@@ -96,6 +96,205 @@ impl Status {
     }
 }
 
+/// Functions applied to an IO stream for receiving input for a UI.
+pub trait UIReader {
+    fn edit<T>(&mut self, contents: &[T]) -> Result<String>
+    where
+        T: fmt::Display;
+    /// Returns true if message reads should expect the source as a tty.
+    fn is_a_tty(&self) -> bool;
+    fn prompt_ask(&mut self, question: &str, default: Option<&str>) -> Result<String>;
+    fn prompt_yes_no(&mut self, question: &str, default: Option<bool>) -> Result<bool>;
+}
+
+/// Functions applied to an IO stream for sending information to a UI.
+pub trait UIWriter {
+    type ProgressBar: DisplayProgress;
+
+    /// IO Stream for sending error messages to.
+    fn err(&mut self) -> &mut io::Write;
+    /// IO Stream for sending normal or informational messages to.
+    fn out(&mut self) -> &mut io::Write;
+    /// Messages sent to the IO streams will be formatted in color if true.
+    fn is_colored(&self) -> bool;
+    /// Messages sent to the IO streams will be formatted for a terminal if true.
+    fn is_a_terminal(&self) -> bool;
+    /// Returns a progress bar widget implementation for writing operation's progress to.
+    fn progress(&self) -> Option<Self::ProgressBar>;
+
+    /// Write a message formatted with `begin`.
+    fn begin<T>(&mut self, message: T) -> io::Result<()>
+    where
+        T: fmt::Display,
+    {
+        let symbol = '»';
```

Comment:
If the only thing different is `symbol`, might it make sense to factor that out to a common `send` function, reducing `begin` and `end` to:
```rust
        self.send(message, '»')
        self.send(message, '★')
```
or take it one step further to and capture the abstraction for `warn` as well:
```rust
        self.send(message, '∅', Self::err)
```

Reply:
I just went with the code that was already there. I think this module could definitely use some more cleanup though 😄

Reply:
My first answer was pretty short but I can see that some more information might be helpful.

This module is a mess. The functions for reading and writing should share much more in common than they currently are, but all of them except for these two were written in a one-off way. `UI` was a struct before and now it's a trait. Implementing a helper function within a trait isn't a good pattern, so I could have opted to move what was `Self::write_heading` to a private module function inside the `ui` module. The abstraction didn't sit well with me in the first place, it followed a rule of 2 vs the rule of 3. When I came through and refactored this I figured it was best to just de-normalize these two functions until somebody had some time to actually dive into the UI module and make some additional improvements. What I have here is just the bare requirements.

If you feel it's necessary I can normalize these two functions into a private module helper function for sure

Reply:
Nah, not a big deal. Thanks for the additional context; I missed the struct/trait distinction.